File: python/tests/test_ontology.py
"""The ontology lookup.

Offline: the endpoint is mocked, so what is checked here is the request that
goes out, the shape that comes back, and that a missing term is an answer
rather than an error.
"""

from __future__ import annotations

import pandas as pd
import pytest
import requests

from seqout.clients.api import SeqoutAPIClient
from seqout.clients.parquet import SeqoutParquetClient
from seqout.exception import SeqoutError

LIVER = {
    "name": "liver",
    "xrefs": ["UBERON:0002107", "MeSH:D008099"],
    "has_children": True,
    "synonyms": [{"name": "iecur", "xrefs": ["UBERON:0002107"]}],
    "synonym_total": 3,
    "children": [
        {"name": "caudate lobe of liver", "has_children": True, "xrefs": ["UBERON:X"]}
    ],
    "children_truncated": False,
    "max_hops": 2,
    "took_ms": 9.7,
}


def _client(payload=LIVER, status=None):
    sq = SeqoutAPIClient()
    seen = {}

    def fake(url, params, response_model):
        seen["url"], seen["params"] = url, params
        if status is not None:
            response = requests.Response()
            response.status_code = status
            raise requests.HTTPError(response=response)
        return response_model.model_validate(payload)

    sq._sender = fake
    return sq, seen


class TestOntology:
    def test_it_reads_the_term_its_synonyms_and_its_children(self):
        sq, _ = _client()
        term = sq.ontology("Liver")
        assert term.name == "liver"
        assert term.xrefs == ["UBERON:0002107", "MeSH:D008099"]
        # synonym_total is the true count; the list is capped by the server.
        assert term.synonym_total == 3
        assert len(term.synonyms) == 1
        assert term.children[0].has_children is True

    def test_sources_names_the_ontologies_behind_the_ids(self):
        sq, _ = _client()
        # CVCL_0030 is a CURIE without a colon, so the prefix has to be read
        # from either separator.
        term = sq.ontology("liver")
        term.xrefs = [*term.xrefs, "CVCL_0030"]
        assert term.sources == ["CVCL", "MeSH", "UBERON"]

    def test_it_asks_the_endpoint_for_what_it_was_given(self):
        sq, seen = _client()
        sq.ontology("liver", 3, children=False)
        assert seen["url"].endswith("/ontology/term")
        assert seen["params"] == {"term": "liver", "max_hops": 3, "children": False}

    def test_a_term_the_graph_lacks_is_none_not_an_error(self):
        sq, _ = _client(status=404)
        assert sq.ontology("zzznope") is None

    def test_any_other_failure_still_raises(self):
        sq, _ = _client(status=500)
        with pytest.raises(requests.HTTPError):
            sq.ontology("liver")

    def test_the_parquet_backend_says_it_has_no_graph(self):
        with pytest.raises(SeqoutError, match="ontology graph"):
            SeqoutParquetClient().ontology("liver")


# "t cell" carries its own CL id; "nbc" carries none and has to borrow from a
# synonym one hop away.
GRAPH = {
    "t cell": {
        "name": "t cell",
        "xrefs": ["CL:0000084", "MeSH:D013601"],
        "synonyms": [{"name": "immature t cell", "xrefs": ["CL:0002420"]}],
    },
    "nbc": {
        "name": "nbc",
        "xrefs": [],
        "synonyms": [{"name": "naive b cell", "xrefs": ["CL:0000788"]}],
    },
}


def _graph_client():
    sq = SeqoutAPIClient()
    asked = []

    def fake(url, params, response_model):
        asked.append(params["term"])
        payload = GRAPH.get(params["term"].lower())
        if payload is None:
            response = requests.Response()
            response.status_code = 404
            raise requests.HTTPError(response=response)
        return response_model.model_validate(payload)

    sq._sender = fake
    return sq, asked


class TestMapToOntology:
    def _frame(self):
        return pd.DataFrame({"celltype": ["T cell", "nbc", "T cell", None, "zzz"]})

    def test_it_adds_the_ids_beside_the_column_it_was_given(self):
        sq, asked = _graph_client()
        out = sq.map_to_ontology(self._frame(), "celltype", use_synonyms=True)
        ids = out["celltype_ontology_id"].tolist()

        assert ids[0] == "CL:0000084,MeSH:D013601"
        # A label with no id of its own borrows from one hop away.
        assert ids[1] == "CL:0000788"
        assert ids[2] == ids[0]
        # A blank cell and a label the graph lacks are both NA, not an error.
        assert out["celltype_ontology_id"].isna().tolist() == [
            False,
            False,
            False,
            True,
            True,
        ]
        # One request per distinct label, however many rows repeat it.
        assert sorted(asked) == ["T cell", "nbc", "zzz"]

    def test_a_synonym_never_overrides_an_id_the_label_has_itself(self):
        # The synonym edge joins a narrower concept, so borrowing CL:0002420
        # here would map "T cell" to immature T cells.
        sq, _ = _graph_client()
        out = sq.map_to_ontology(self._frame(), "celltype", use_synonyms=True)
        assert "CL:0002420" not in out["celltype_ontology_id"][0]

    def test_synonyms_are_off_until_they_are_asked_for(self):
        sq, _ = _graph_client()
        ids = sq.map_to_ontology(self._frame(), "celltype")["celltype_ontology_id"]
        # "nbc" carries nothing of its own, so by default it stays unmapped
        # rather than taking the identifier of a term one hop away.
        assert ids[0] == "CL:0000084,MeSH:D013601"
        assert pd.isna(ids[1])

    def test_one_ontology_can_be_asked_for_on_its_own(self):
        sq, _ = _graph_client()
        out = sq.map_to_ontology(
            self._frame(), ["celltype"], ontology="MeSH", use_synonyms=True
        )
        assert out["celltype_ontology_id"][0] == "MeSH:D013601"
        assert out["celltype_ontology_id"].isna()[1]  # CL only

    def test_a_column_that_is_not_there_says_so(self):
        sq, _ = _graph_client()
        with pytest.raises(KeyError, match="nope"):
            sq.map_to_ontology(self._frame(), "nope")

File: R/tests/testthat/test-ontology.R
liver_response <- list(
  name = "liver",
  xrefs = list("UBERON:0002107", "MeSH:D008099"),
  has_children = TRUE,
  synonyms = list(
    list(name = "iecur", xrefs = list("UBERON:0002107")),
    list(name = "hepatic", xrefs = list())
  ),
  synonym_total = 3L,
  children = list(
    list(
      name = "caudate lobe of liver",
      has_children = TRUE,
      xrefs = list("UBERON:0001117")
    )
  ),
  children_truncated = FALSE,
  max_hops = 2L
)

onto_rows <- function(res) {
  c(
    list(seqout:::.onto_row(res, "term")),
    lapply(seqout:::.as_record_list(res$synonyms), seqout:::.onto_row,
      relation = "synonym"
    ),
    lapply(seqout:::.as_record_list(res$children), seqout:::.onto_row,
      relation = "child"
    )
  )
}

test_that("the term, its synonyms and its children are one tibble", {
  out <- seqout:::.onto_tibble(onto_rows(liver_response))
  expect_s3_class(out, "tbl_df")
  expect_equal(out$relation, c("term", "synonym", "synonym", "child"))
  expect_equal(out$name[1], "liver")
  expect_equal(out$has_children, c(TRUE, FALSE, FALSE, TRUE))
})

test_that("xrefs is a list column, so a term can carry several", {
  out <- seqout:::.onto_tibble(onto_rows(liver_response))
  expect_type(out$xrefs, "list")
  expect_equal(out$xrefs[[1]], c("UBERON:0002107", "MeSH:D008099"))
  # A term with no identifier keeps an empty vector, not NA.
  expect_equal(out$xrefs[[3]], character(0))
})

test_that("a term the graph lacks is an empty tibble, not an error", {
  out <- seqout:::.onto_tibble(list())
  expect_s3_class(out, "tbl_df")
  expect_equal(nrow(out), 0L)
  expect_equal(names(out), c("relation", "name", "xrefs", "has_children"))
})

test_that("ontology refuses a Parquet connection", {
  expect_error(
    ontology("liver", con = fake_con()),
    "reads the REST API"
  )
})

# "t cell" carries its own CL id; "nbc" carries none and has to borrow from a
# synonym one hop away.
graph_tibble <- function(term) {
  rows <- list(
    "t cell" = list(
      seqout:::.onto_row(
        list(name = "t cell", xrefs = list("CL:0000084", "MeSH:D013601")), "term"
      ),
      seqout:::.onto_row(
        list(name = "immature t cell", xrefs = list("CL:0002420")), "synonym"
      )
    ),
    "nbc" = list(
      seqout:::.onto_row(list(name = "nbc", xrefs = list()), "term"),
      seqout:::.onto_row(
        list(name = "naive b cell", xrefs = list("CL:0000788")), "synonym"
      )
    )
  )
  found <- rows[[tolower(term)]]
  seqout:::.onto_tibble(if (is.null(found)) list() else found)
}

test_that("a label takes its own identifiers, and a synonym's only when asked", {
  own <- seqout:::.ontology_ids(graph_tibble("t cell"), use_synonyms = TRUE)
  # The synonym link joins a narrower concept, so CL:0002420 must not ride
  # along while the label has an identifier of its own.
  expect_equal(own, "CL:0000084,MeSH:D013601")

  # "nbc" carries nothing itself: unmapped by default, mapped when the
  # synonyms are asked for.
  expect_true(is.na(seqout:::.ontology_ids(graph_tibble("nbc"))))
  expect_equal(
    seqout:::.ontology_ids(graph_tibble("nbc"), use_synonyms = TRUE),
    "CL:0000788"
  )
  expect_true(is.na(seqout:::.ontology_ids(graph_tibble("zzz"))))
})

test_that("one ontology can be asked for on its own", {
  expect_equal(seqout:::.ontology_ids(graph_tibble("t cell"), "MeSH"), "MeSH:D013601")
  # CVCL_0030 has no colon, so the prefix has to be read from either separator.
  hela <- seqout:::.onto_tibble(list(
    seqout:::.onto_row(list(name = "hela", xrefs = list("CVCL_0030", "EFO:0001185")), "term")
  ))
  expect_equal(seqout:::.ontology_ids(hela, "CVCL"), "CVCL_0030")
  expect_true(is.na(seqout:::.ontology_ids(hela, "CL")))
})

test_that("map_to_ontology adds one column per column, and refuses Parquet", {
  testthat::local_mocked_bindings(ontology = function(term, ...) graph_tibble(term))
  meta <- data.frame(
    celltype = c("T cell", "nbc", "T cell", NA), n = 1:4,
    stringsAsFactors = FALSE
  )

  out <- map_to_ontology(meta, "celltype", use_synonyms = TRUE)
  expect_equal(names(out), c("celltype", "n", "celltype_ontology_id"))
  expect_equal(
    out$celltype_ontology_id,
    c("CL:0000084,MeSH:D013601", "CL:0000788", "CL:0000084,MeSH:D013601", NA)
  )

  # Off by default: "nbc" has no identifier of its own to give.
  expect_equal(
    map_to_ontology(meta, "celltype")$celltype_ontology_id,
    c("CL:0000084,MeSH:D013601", NA, "CL:0000084,MeSH:D013601", NA)
  )

  expect_error(map_to_ontology(meta, "nope"), "not a column")
  expect_error(map_to_ontology(meta, "celltype", con = fake_con()), "REST API")
})
